pow/src/main/java/tech/pegasys/teku/pow/Eth1BlockFetcher.java

```diff
@@ -0,0 +1,153 @@
+    if (active) {
+      return;
+    }
+    active = true;
+    LOG.debug("Beginning back-fill of Eth1 blocks");
+    backfillEth1Blocks(latestCanonicalBlockNumber);
+  }
+
+  public void fetch(final BigInteger fromBlock, final BigInteger toBlock) {
+    synchronized (this) {
+      if (!active) {
+        // No point starting to request data if we haven't yet reached the start of the chain
+        // We'll just wind up downloading a heap of blocks we then throw away
+        return;
+      }
+      for (BigInteger block = fromBlock;
+          block.compareTo(toBlock) <= 0;
+          block = block.add(BigInteger.ONE)) {
+        blocksToRequest.add(UnsignedLong.valueOf(block));
+      }
+    }
+    requestNextBlockIfRequired();
+  }
+
+  private void requestNextBlockIfRequired() {
+    UnsignedLong blockToRequest;
+    synchronized (this) {
+      if (requestInProgress || blocksToRequest.isEmpty()) {
+        return;
+      }
+      blockToRequest = blocksToRequest.last();
+      blocksToRequest.remove(blockToRequest);
+      requestInProgress = true;
+    }
+    requestBlock(blockToRequest)
+        .always(
+            () -> {
+              synchronized (Eth1BlockFetcher.this) {
+                requestInProgress = false;
+              }
+              requestNextBlockIfRequired();
+            });
+  }
+
+  private SafeFuture<Void> requestBlock(final UnsignedLong blockNumberToRequest) {
+    // Note: Not using guaranteed requests here - if the Eth1 chain is temporarily unavailable
+    // we may miss some blocks but that's better than potentially getting stuck retrying a block
+    LOG.debug("Requesting block {}", blockNumberToRequest);
+    return eth1Provider
+        .getEth1BlockFuture(blockNumberToRequest)
+        .thenAccept(
+            block -> {
+              if (isAboveLowerBound(UnsignedLong.valueOf(block.getTimestamp()))) {
+                postBlock(block);
+              } else {
+                // Every block before the one we just fetched must be outside of the range
+                synchronized (Eth1BlockFetcher.this) {
+                  // All blocks at or before this number must be before the cache period
+                  blocksToRequest.headSet(blockNumberToRequest, true).clear();
+                }
+              }
+              requestNextBlockIfRequired();
+            })
+        .exceptionallyCompose(
+            error -> {
+              LOG.warn("Failed to retrieve block {}", blockNumberToRequest);
+              return SafeFuture.COMPLETE;
+            });
+  }
+
+  private void postBlock(final Block block) {
+    eth1EventsChannel.onEth1Block(
+        Bytes32.fromHexString(block.getHash()), UnsignedLong.valueOf(block.getTimestamp()));
+  }
+
+  private void backfillEth1Blocks(final UnsignedLong nextBlockToRequest) {
+    // Walk backwards from blockNumber until we reach the start of the voting period
+    eth1Provider
+        .getGuaranteedEth1BlockFuture(nextBlockToRequest)
```

Comment:
Is there a reason why you used guaranteed requests here? Missing some blocks does seem better than potentially getting stuck retrying a block, similar to the case above.

Reply:
I'm in two minds about it.  Ultimately I went with guaranteed here in backfill and not the normal case because if the node is down or overloaded briefly during backfill we'll potentially miss fairly large numbers of blocks and not backfill the cache properly.  If it's down briefly in the normal case, we likely only miss a couple of blocks (because normally we're querying one new block about every 15 seconds).

Reply:
That makes sense to me.